src/settings/README.md

```diff
@@ -0,0 +1,90 @@
+
+### Internal Map
+
+Although there is a physical `pg_settings` table in the catalogs where we can find all the information about a parameter, we finally decided that there should be a map that contains the values of parameters. Why? Because:
+
+- The most important reason is that accessing the catalog needs to be done in a transaction. This can cause infinite loops. A typical situation is, on system bootstrapping, the Settings Manager needs to open a transaction to append the values into the `pg_settings` table. However, as long as there is a configuration needed by the Transaction Manager (a very likely one can be the default isolation level), it will call `get` to ask for that parameter, and the Settings Manager will open a new transaction to read the parameter from the catalog, which results in an infinite loop. In order to reduce the interaction with other parts of the system, we decided to maintain a map inside the Settings Manager. In this way, it can serve all requests without consulting anyone else.
+- It is much easier to operate on a map than the catalog. 
+- Since the number of parameters is low, an extra map won't introduce much overhead.
+
+- Although this essentially turned the Settings Manager as a cache layer for the `pg_settings` table, its consistency is easy to maintain (just update the catalog in `set` methods).
+
+## Testing Plan
+
+> How should the component be tested?
+
+The following tests should be written:
+
+- Are the default values loaded successfully?
+- Do `gflags` and config file successfully overwrite default values?
+- Can mutable parameters be correctly changed? If it has a callback function, is that called?
+- Can the Settings Manager check if the new value is between `min_value` and `max_value`?
+- Can the Settings Manager refuse to change immutable parameters?
+
+## Trade-offs and Potential Problems
+
+> Write down any conscious trade-off you made that can be problematic in the future or any problems discovered during the design process that remain unaddressed (technical debts).
+
+By adding the internal map, the system is storing 3 copies of each parameter value: one in its component (each component is very likely to maintain a local copy, like in the Garbage Collector there is very likely to be a `num_threads` member variable), one in the map, and one in the catalog. However, this is necessary as stated above. The real unnecessary copy is actually the `pg_settings` table. In fact, Postgresql does not have this physical table. It also stores the values in an internal map. When someone calls `SELECT * FROM pg_settings`, it generates a view from the internal map. However, Terrier does not support generating a view from functions yet. Therefore, we must maintain the physical table.
```

Comment:
Since there are 3 copies of each parameter value, will those values come into an inconsistent state? If so, how do you plan to handle this inconsistency?

Reply:
No, as long as the component is using our setters and getters to change and load configurations.